### ISSUE TITLE
nats ca rotation

### ISSUE THREAD

**Issue by glupu** [cloudfoundry/bosh-deployment#334]
Hello,

We're trying to follow https://bosh.io/docs/nats-ca-rotation/ in order to renew the expired certificates. After the first step, when we introduced the new certificate as per the documentation, everything went south. All vms in the deployments are showing as "unresponsive agent", ssh doesn't work.

I tried recreate on one of the deployments, I can see the vms as active in the openstack, however, bosh will keep pinging until it timeouts after 10 minutes. If I try to connect to 22 from the director to the vms is working fine, so clearly its not a network issue.

WHen I try to bosh ssh one of the unresponsive vms I get this in the nats log on director:

[22917] 2019/02/28 18:08:51.212998 [ERR] 10.110.147.1:55386 - cid:202618 - An error occurred while peeking connection: read tcp 10.110.147.1:4222->10.110.147.1:55386: read: connection reset by peer
[22917] 2019/02/28 18:08:51.213019 [ERR] 10.110.147.1:55386 - cid:202618 - An error occurred while detecting if client requested TLS: Expected header size to be 6, but was 0

If someone got any idea will be much appreciated.

Thanks,
George

**Comment by h4xnoodle** [comment]
Hello @glupu,

Were your certificates already expired? Were the VMs already unresponsive?

In order to try to fix those unresponsive VMs, try `bosh -d <deployment> recreate --fix`. "Fix" will ignore the unresponsive state. 

~ Rebecca (Bosh, Toronto)

**Comment by glupu** [comment]
Hello Rebecca,

Thanks a lot for your response.

All the certificates were expired (we couldn't do any bosh commands, like deployments or vms), so we don't know if the VMs were in "unresponsive status" before.

We removed all the certificates except the nats one (for which we are following the procedure) and recreated the bosh env. After, we were able to do bosh deployments or vms, however, all the VMs are in unresponsive agent status. Also, we noticed lots of queued tasks, 100-200 (scan and fix). We  stopped health_monitor process on the director and cancel all of them.

Current situation is that we cannot recreate a deployment, as we tried with a concourse one:

Task 10962 | 06:20:57 | Preparing deployment: Preparing deployment (00:00:01)
Task 10962 | 06:20:59 | Preparing package compilation: Finding packages to compile (00:00:00)
Task 10962 | 06:20:59 | Creating missing vms: web/98e0f515-9194-43a0-8605-75757084c7aa (0)
Task 10962 | 06:20:59 | Creating missing vms: db/a9cc5fc9-5793-4434-b84d-5e637094f190 (0)
Task 10962 | 06:20:59 | Creating missing vms: worker/50536c97-a969-489b-bf22-9ac0b71d599f (0)
Task 10962 | 06:31:33 | Creating missing vms: web/98e0f515-9194-43a0-8605-75757084c7aa (0) (00:10:34)
                     L Error: Timed out pinging to 14036d94-041d-4ebe-b8d2-b1789e7a5bee after 600 seconds
Task 10962 | 06:31:33 | Creating missing vms: db/a9cc5fc9-5793-4434-b84d-5e637094f190 (0) (00:10:34)
                     L Error: Timed out pinging to 7cf8fadb-f6e7-4e34-82e1-b440655d4a7b after 600 seconds
Task 10962 | 06:31:33 | Creating missing vms: worker/50536c97-a969-489b-bf22-9ac0b71d599f (0) (00:10:34)
                     L Error: Timed out pinging to 3c227b05-39f7-4a0c-a9b9-3f5f52498a06 after 600 seconds
Task 10962 | 06:31:33 | Error: Timed out pinging to 14036d94-041d-4ebe-b8d2-b1789e7a5bee after 600 seconds

Task 10962 Started  Fri Mar  1 06:20:57 UTC 2019
Task 10962 Finished Fri Mar  1 06:31:33 UTC 2019
Task 10962 Duration 00:10:36
Task 10962 error

Changing state:
  Expected task '10962' to succeed but state is 'error'

Exit code 1

VMs are only accessible if I ssh directly (not thru bosh), but I'm limited in terms of logs, as I can only login as vcap.

The other deployments/vms are still there and working fine, however, we cannot "manage" them with this director.

One weird thing to note is that we have one deployment with one VM (its a concourse single vm deployment) that its showing fine and we could connect with bosh ssh on it.

Any help would be much appreciated.

George

**Comment by h4xnoodle** [comment]
What are the commands you're running to recreate the deployment? You will also need `--fix`. Fix ignores the connectivity problems and doesn't attempt to communicate via NATS. https://bosh.io/docs/cli-v2/#recreate 

The resurrector is also trying to repair the VMs using the new NATS credentials. The scan and fix tasks you see is the resurrector kicking in to recover. It may be best to turn off the resurrector for now and cancel the scan & fix tasks. `bosh update-resurrection off`. Be sure to turn it back on after.

Unfortunately for the case where NATS certs have already expired, you will see unresponsive VMs everywhere. The supported method to repair these deployments is to `recreate` with `--fix`.

Hope this helps!

**Comment by glupu** [comment]
The problem here is that even if I use --fix it still complains about NATS certificates. Let me explain:

1\ I'm recreating the vms with recreate --fix
2\ Bosh will say this:
Using environment '10.110.147.1' as client 'admin'

Task 11112

Task 11112 | 10:26:04 | Preparing deployment: Preparing deployment (00:00:00)
Task 11112 | 10:26:05 | Preparing package compilation: Finding packages to compile (00:00:00)
Task 11112 | 10:26:05 | Creating missing vms: web/98e0f515-9194-43a0-8605-75757084c7aa (0)
Task 11112 | 10:26:05 | Creating missing vms: db/a9cc5fc9-5793-4434-b84d-5e637094f190 (0)
Task 11112 | 10:26:05 | Creating missing vms: worker/50536c97-a969-489b-bf22-9ac0b71d599f (0)
Task 11112 | 10:36:52 | Creating missing vms: web/98e0f515-9194-43a0-8605-75757084c7aa (0) (00:10:47)
                     L Error: Timed out pinging to 6440cc9f-837a-4241-ba2d-a75ca9da9595 after 600 seconds
Task 11112 | 10:36:53 | Creating missing vms: db/a9cc5fc9-5793-4434-b84d-5e637094f190 (0) (00:10:48)
                     L Error: Timed out pinging to 8171cf69-a64d-46f6-b93f-6d25e65df244 after 600 seconds
Task 11112 | 10:36:57 | Creating missing vms: worker/50536c97-a969-489b-bf22-9ac0b71d599f (0) (00:10:52)
                     L Error: Timed out pinging to 2ca13623-2abb-4ee3-b29a-3fb91704bf4c after 600 seconds
Task 11112 | 10:36:57 | Error: Timed out pinging to 6440cc9f-837a-4241-ba2d-a75ca9da9595 after 600 seconds

Task 11112 Started  Tue Mar  5 10:26:04 UTC 2019
Task 11112 Finished Tue Mar  5 10:36:57 UTC 2019
Task 11112 Duration 00:10:53
Task 11112 error

Capturing task '11112' output:
  Expected task '11112' to succeed but state is 'error'

Exit code 1

In debug it just does this for 10minutes:

D, [2019-03-05T10:33:22 #27218] [create_missing_vm(db/a9cc5fc9-5793-4434-b84d-5e637094f190 (0)/3)] DEBUG -- DirectorJobRunner: (0.000294s) SELECT * FROM "tasks" WHERE "id" = 11112
D, [2019-03-05T10:33:22 #27218] [create_missing_vm(db/a9cc5fc9-5793-4434-b84d-5e637094f190 (0)/3)] DEBUG -- DirectorJobRunner: SENT: agent.8171cf69-a64d-46f6-b93f-6d25e65df244 {"protocol":3,"method":"ping","arguments":[],"reply_to":"director.4e27310a-14d6-4592-bb22-e742844be395.8171cf69-a64d-46f6-b93f-6d25e65df244.707680d3-ba65-4946-b374-2cd2509cd7a7"}
D, [2019-03-05T10:33:23 #27218] [create_missing_vm(web/98e0f515-9194-43a0-8605-75757084c7aa (0)/3)] DEBUG -- DirectorJobRunner: (0.000251s) SELECT * FROM "tasks" WHERE "id" = 11112
D, [2019-03-05T10:33:23 #27218] [create_missing_vm(web/98e0f515-9194-43a0-8605-75757084c7aa (0)/3)] DEBUG -- DirectorJobRunner: SENT: agent.6440cc9f-837a-4241-ba2d-a75ca9da9595 {"protocol":3,"method":"ping","arguments":[],"reply_to":"director.4e27310a-14d6-4592-bb22-e742844be395.6440cc9f-837a-4241-ba2d-a75ca9da9595.751e0449-3ff3-4002-82b3-de192485cc42"}
D, [2019-03-05T10:33:23 #27218] [create_missing_vm(worker/50536c97-a969-489b-bf22-9ac0b71d599f (0)/3)] DEBUG -- DirectorJobRunner: (0.000328s) SELECT * FROM "tasks" WHERE "id" = 11112
D, [2019-03-05T10:33:23 #27218] [create_missing_vm(worker/50536c97-a969-489b-bf22-9ac0b71d599f (0)/3)] DEBUG -- DirectorJobRunner: SENT: agent.2ca13623-2abb-4ee3-b29a-3fb91704bf4c {"protocol":3,"method":"ping","arguments":[],"reply_to":"director.4e27310a-14d6-4592-bb22-e742844be395.2ca13623-2abb-4ee3-b29a-3fb91704bf4c.2bfb2dd3-fd5e-4a56-a23e-5815c71adea5"}

Now, what I did was to nova rescue one of the VMs and mount the rootpartition and check the /var/vcap/bosh/log/current log file and I have this:

2019-03-05_10:33:02.68388 [attemptRetryStrategy] 2019/03/05 10:33:02 DEBUG - Making attempt #2 for *retrystrategy.retryable
2019-03-05_10:33:02.68391 [Cmd Runner] 2019/03/05 10:33:02 DEBUG - Running command 'arp -d 10.110.147.1'
2019-03-05_10:33:02.68500 [Cmd Runner] 2019/03/05 10:33:02 DEBUG - Stdout:
2019-03-05_10:33:02.68501 [Cmd Runner] 2019/03/05 10:33:02 DEBUG - Stderr:
2019-03-05_10:33:02.68501 [Cmd Runner] 2019/03/05 10:33:02 DEBUG - Successful: true (0)
2019-03-05_10:33:03.69795 [attemptRetryStrategy] 2019/03/05 10:33:03 DEBUG - Making attempt #3 for *retrystrategy.retryable
2019-03-05_10:33:03.69803 [Cmd Runner] 2019/03/05 10:33:03 DEBUG - Running command 'arp -d 10.110.147.1'
2019-03-05_10:33:03.69914 [Cmd Runner] 2019/03/05 10:33:03 DEBUG - Stdout:
2019-03-05_10:33:03.69917 [Cmd Runner] 2019/03/05 10:33:03 DEBUG - Stderr:
2019-03-05_10:33:03.69919 [Cmd Runner] 2019/03/05 10:33:03 DEBUG - Successful: true (0)
2019-03-05_10:33:03.80598 [Cmd Runner] 2019/03/05 10:33:03 DEBUG - Running command 'arping -c 1 -U -I eth0 10.110.147.102'
2019-03-05_10:33:03.82542 [Cmd Runner] 2019/03/05 10:33:03 DEBUG - Stdout: ARPING 10.110.147.102 from 10.110.147.102 eth0
2019-03-05_10:33:03.82543 Sent 1 probes (1 broadcast(s))
2019-03-05_10:33:03.82544 Received 0 response(s)
2019-03-05_10:33:03.82544 [Cmd Runner] 2019/03/05 10:33:03 DEBUG - Stderr:
2019-03-05_10:33:03.82544 [Cmd Runner] 2019/03/05 10:33:03 DEBUG - Successful: true (0)
2019-03-05_10:33:04.71226 [main] 2019/03/05 10:33:04 ERROR - App run Running agent: Message Bus Handler: Starting nats handler: Connecting: Connecting to NATS: x509: certificate has expired or is not yet valid
2019-03-05_10:33:04.71230 [main] 2019/03/05 10:33:04 ERROR - Agent exited with error: Running agent: Message Bus Handler: Starting nats handler: Connecting: Connecting to NATS: x509: certificate has expired or is not yet valid
2019-03-05_10:33:04.72157 [main] 2019/03/05 10:33:04 DEBUG - Starting agent


so it just does the arping and then throw the "certificate has expired" and try again, for 10 minutes

If my understanding is correct, the --fix doesn't do what it's supposed to do?

**Comment by h4xnoodle** [comment]
It looks like you're recreating with the same expired certificates as before.

Can you please paste *exactly* what you're running? The correct strategy should be the following when the certificates *are already expired*:

1. Remove the `nats_*` *keys and values* from `creds.yml`. The CLI will then regenerate all of these for you.
2. `bosh create-env ...` to update the director to use the new certificates
3. `bosh recreate -d <deployment> --fix`

Have you updated your director prior to recreating?

**Comment by glupu** [comment]
Yep. What I did was to follow the documentation here: https://bosh.io/docs/nats-ca-rotation/ . At first step, I tried to do a create-env using a new ops file that introduces a new CA and concatenates the old and the new. Afterwards, I was trying to recreate my VMs, so I can introduce the new CA to them also, however, this is where I was stuck.

I was able to fix my problem by following what h4xnoodle suggested, to remove all the certificates (nats) from the creds yml file and create-env the bosh. Then, recreate --fix worked fine.

Thanks a lot for your help!

George
